### ISSUE TITLE
Misleading ExpressionLanguage Documentation

### ISSUE THREAD

**Issue by dbohannon** [symfony/symfony-docs#8259]
The ExpressionLanguage component page (https://symfony.com/doc/current/components/expression_language.html) states:

> “Expressions can be seen as a very restricted PHP sandbox and are immune to external injections as you must explicitly declare which variables are available in an expression.” 

However, this statement is a bit misleading as expression injection is very possible when mapping tainted data to expression variables in the $values array. Even though an injected expression is limited to a predefined set of variables and operators, an attacker can still modify the expression and alter application logic. The documentation also states that the ExpressionLanguage component may be used for validation which further increases the likelihood that a developer will pass tainted data into the expression and create an expression injection vulnerability. I suggest updating the documentation at https://github.com/symfony/symfony-docs/blob/master/components/expression_language.rst to explicitly state that tainted data should not be passed to the $values array (or appended directly to the expression string). 

To demonstrate the issue, consider the example code block below:
```PHP
$user = $_GET['user'];
if(!$language->evaluate('user["username"] matches "/[a-zA-Z0-9]$/"', array("user" => $user))){
 return new Response("Validation failed!");
}
```
Passing the following user parameter will alter the expression and bypass the validation:
`<script>alert('XSS_SUCCESS!')</script>" > 0 | "a`

I will try to update the documentation and submit a pull request as time permits.

**Comment by javiereguiluz** [comment]
@dbohannon this is an interesting proposal. Thanks for the insights!

However, instead of adding more explanations or complicating docs, what if we just remove this controversial paragraph entirely?

**Comment by dbohannon** [comment]
I still think the documentation should explicitly state that tainted data should not be passed to the $values array, especially since other documentation pages talk about using ExpressionLanguage for validation. If a developer attempts to validate user-controlled data then they will be open to expression injection.

**Comment by weaverryan** [comment]
Yea, we do need a warning *somewhere* about this - it's not obvious (at least to me) that a dynamic value has the ability to affect the underlying logic like this.

@dbohannon what other examples are there where this might be used? I can see with your example that you could "fool" the value being passed to `matches`. I suppose you could also do with pretty much any operand?

```
$price = $_GET['price'];
if(!$language->evaluate('price < 0', array("price" => $price))){
 return new Response("Validation failed!");
}
```

being passed the following (string) value:

```
-100 + 200
```

Is that another example? I'm trying to see how generic/specific the problem is to determine the best wording. Also, is there a generic way to escape the input values, or not? My instinct is no - it's not a matter of escaping certain characters... it depends on how you're using the value, correct?

Thanks!

**Comment by dbohannon** [comment]
You're correct that you can use pretty much any built-in operand to alter the expression logic. The operands used will vary depending on the implementation and how you're trying to exploit it. I believe that you would have to sanitize or escape all operands listed at http://symfony.com/doc/current/components/expression_language/syntax.html as well as ', ", etc. Or, maybe just white-list alphanumeric characters. So maybe the recommendation should be "Avoid passing untrusted data into the $values array. If untrusted data must be used within the expression, sanitize non-alphanumeric characters to prevent malicious users from injecting control characters and altering the expression". 

Another possibility is to rework the ExpressionLanguage implementation so that it parameterizes the expression and then inserts data from the $values array (similar to a parameterized SQL query), but I'm not sure how feasible this is since it would probably require a bit of work and be a breaking change.

**Comment by javiereguiluz** [comment]
Closing as fixed by #13688.

**Comment by valepu** [comment]
Hi,

I'm writing on this old issue because I think there's either been a mistake or some input sanitization was added to the Expression Language in a later version (I'm currently using 6.2.2).

Today while reading ExpressionLanguage documentation I stumbled upon the warning added due to this issue and was trying to find a way to properly escape the variables but ended up finding this issue. I was getting ready to test this myself but I noticed the regex used in the first example is wrong.

The regex `"/[a-zA-Z0-9]$/"` used in the first post will test if the last character of the string is alphanumeric, which is the case for `<script>alert('XSS_SUCCESS!')</script>" > 0 | "a` so the expression is returning true because the last character of the string is "a" not because the expression was altered through the variable. You can test it here: https://onlinephp.io?s=s7EvyCjg5eLlSk3OyFcoKEpNj89NLEnO0FDSj07UrXLUjTLQtYxV0VfSUVCyKU4uyiwosUvMSS0q0VCPCA6ODw51dnYNDlZU17TRh8rGKCnYKRgo1CjEKCUqaSrYKyj5ggxMLVZSsFJQcslPLc5TL1EAW6JkDQA%2C&v=7.4.30

I tested this example using this regex  `"/^[a-zA-Z0-9]+$/"` (which will return true only if the whole string is alphanumeric) and this time I got the proper answer (matching `<script>alert('XSS_SUCCESS!')</script>" > 0 | "a` against `"/^[a-zA-Z0-9]+$/"` returned false).

I also tried these: 
- `$expr->evaluate("value", ["value" => '100 - 1']);` returned the string "100 - 1"
- `$expr->evaluate("value + 1", ["value" => '100 - 1']);` gave me error (A non-numeric value encountered)

  
- `$expr->evaluate("value < 0", ["value" => '50 - 100']);` returned false
- `$expr->evaluate("value > 0", ["value" => '50 - 100']);` returned true. But I believe it's converting the string "50 - 100" into "50" (at least that's what `intval()` returns) just like normal PHP would do

While I'm all for telling people to sanitize inputs as often as possible I am wondering if this warning was added improperly and there's actually no need to sanitize inputs used in the expression language beyond what is expected for other kinds of inputs
